lib/web/addr.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package web
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+	"net/netip"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	// xForwardedForHeader is a de-facto standard header for identifying the
+	// originating IP address of a client connecting to a web server through a
+	// proxy server.
+	xForwardedForHeader = "X-Forwarded-For"
+)
+
+// maybeUpdateClientSrcAddr overwrites client source address if X-Forwarded-For
+// is set.
+//
+// Both hijacked conn and request context are updated. The hijacked conn can be
+// used for ALPN connection upgrades or Websocket connections.
+func maybeUpdateClientSrcAddr(w http.ResponseWriter, r *http.Request) (http.ResponseWriter, *http.Request) {
+	// multiplexer handles extracting real client IP using PROXY protocol where
+	// available, so we can omit checking X-Forwarded-For.
+	if hasClientAddrFromMultiplexer(r.Context()) {
+		return w, r
+	}
+
+	// Use X-Forwarded-for if set.
+	xForwardedFor := r.Header.Get(xForwardedForHeader)
+	if xForwardedFor == "" {
+		return w, r
+	}
+
+	forwardedAddr, err := getForwardedAddr(r.RemoteAddr, xForwardedFor)
+	if err != nil {
+		logrus.Debugf("Invalid X-Forwarded-For %q: %v.", xForwardedFor, err)
+		return w, r
+	}
```

Comment:
> I think that before doing that we must ensure that this is the Connection Upgrade flow otherwise the X-Forwarded-For header can be set on a user side to hide/cover their real IP address and access web teleport service protected by IP pinning like Teleport Desktop Access /webapi/sites/:site/desktops/:desktopName/connect

I believe `X-Forwarded-For` should be used for all requests.

First, the login APIs must take addresses from `X-Forwarded-For` otherwise the certificate will use load balancers' ips. Same for logins from either tsh or the web UI.

The websocket for SSH or Desktop on web UI should also take  `X-Forwarded-For` to match the login. Without this change, WebUI behind l7 lb randomly fails (without PROXY protocol) as sometimes you get lucky with same load balancer ip but other times it fails. 

An alternative to the current approach is for `tsh` to mark all login APIs with a flag something like `teleport-x-use-x-forwarded-for: true` so the server knows when to take header for logins. I started with this solution and it worked great for `tsh`  but I couldn't find a way for the WebUI to do the "handshake` test to detect if it's behind a l7 loadbalancer.

Let me know what you think. Thanks!